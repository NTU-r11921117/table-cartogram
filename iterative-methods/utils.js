/* eslint-disable complexity */
export function translateVectorToTable(vector, targetTable, height, width) {
  // vector index tracks the position in the vector as the double loop moves across
  // the form of the output table
  let vectorIndex = 0;
  const newTable = [];
  for (let i = 0; i <= targetTable.length; i++) {
    const newRow = [];
    for (let j = 0; j <= targetTable[0].length; j++) {
      const inFirstRow = i === 0;
      const inLeftColumn = j === 0;
      const inRightColumn = j === targetTable[0].length;
      const inLastRow = i === targetTable.length;
      const inCorner = ((inFirstRow && (inLeftColumn || inRightColumn))) ||
              ((inLastRow && (inLeftColumn || inRightColumn)));
      if (inCorner) {
        newRow.push({x: inLeftColumn ? 0 : width, y: inFirstRow ? 0 : height});
      } else if (inFirstRow || inLastRow) {
        newRow.push({x: vector[vectorIndex], y: inFirstRow ? 0 : height});
        vectorIndex = vectorIndex + 1;
      } else if (inLeftColumn || inRightColumn) {
        newRow.push({x: inLeftColumn ? 0 : width, y: vector[vectorIndex]});
        vectorIndex = vectorIndex + 1;
      } else {
        newRow.push({x: vector[vectorIndex], y: vector[vectorIndex + 1]});
        vectorIndex = vectorIndex + 2;
      }
    }
    newTable.push(newRow);
  }
  return newTable;
}

export function translateTableToVector(table, targetTable) {
  const vector = [];
  for (let i = 0; i < table.length; i++) {
    for (let j = 0; j < table[0].length; j++) {
      const inFirstRow = i === 0;
      const inLeftColumn = j === 0;
      const inRightColumn = j === targetTable[0].length;
      const inLastRow = i === targetTable.length;
      const cell = table[i][j];
      const inCorner = ((inFirstRow && (inLeftColumn || inRightColumn))) ||
              ((inLastRow && (inLeftColumn || inRightColumn)));
      if (inCorner) {
        // do nothing
      } else if (inFirstRow || inLastRow) {
        vector.push(cell.x);
      } else if (inLeftColumn || inRightColumn) {
        vector.push(cell.y);
      } else {
        vector.push(cell.x);
        vector.push(cell.y);
      }
    }
  }
  return vector;
}

export function getRectsFromTable(table) {
  const rects = [];
  for (let i = 0; i < table.length - 1; i++) {
    const rowRects = [];
    for (let j = 0; j < table[0].length - 1; j++) {

      rowRects.push([
        table[i][j],
        table[i + 1][j],
        table[i + 1][j + 1],
        table[i][j + 1]
      ]);
    }
    rects.push(rowRects);
  }
  return rects;
}

export function findSumForTable(areas) {
  return areas.reduce((sum, row) => row.reduce((acc, cell) => acc + cell, sum), 0);
}

export function findMaxForTable(areas) {
  return areas.reduce((max, row) => row.reduce((acc, cell) => Math.max(acc, cell), max), -Infinity);
}

export function convertToManyPolygons(table) {
  const rects = [];
  for (let i = 0; i < table.length - 1; i++) {
    for (let j = 0; j < table[0].length - 1; j++) {
      const newRect = [
        table[i][j],
        table[i + 1][j],
        table[i + 1][j + 1],
        table[i][j + 1]
      ];
      rects.push(newRect);
    }
  }
  return rects;
}

const diffVecs = (a, b) => ({x: a.x - b.x, y: a.y - b.y});
const dotVecs = (a, b) => a.x * b.x + a.y * b.y;
const normVec = a => Math.sqrt(Math.pow(a.x, 2) + Math.pow(a.y, 2));
export function checkForConcaveAngles(rect) {
  for (let i = 1; i < 4; i++) {
    const aVec = diffVecs(rect[i], rect[i - 1]);
    const bVec = diffVecs(rect[i], rect[(i + 1) === 4 ? 0 : (i + 1)]);
    const cosVal = dotVecs(aVec, bVec) / (normVec(aVec) * normVec(bVec));
    const angle = Math.acos(cosVal);
    // console.log(angle)
    if (angle >= Math.PI) {
      return true;
      // console.log(rect)
    }
  }
  return false;
}

export function matrixIterate(width, height, cell) {
  const output = [];
  for (let i = 0; i < width; i++) {
    const row = [];
    for (let j = 0; j < height; j++) {
      row.push(cell(j, i));
    }
    output.push(row);
  }
  return output;
}

export const matrixAdd = (matA, matB) =>
  matrixIterate(matA[0].length, matA.length, (x, y) => matA[y][x] + matB[y][x]);

export const transposeMatrix = mat => mat[0].map((col, i) => mat.map(row => row[i]));

export const times = n => [...new Array(n)].map((d, i) => i);
