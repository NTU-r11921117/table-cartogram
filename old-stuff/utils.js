const pointAdd = (l, r) => (l.x * r.y - r.x * l.y);

/**
 * Generate a number of equally spaced points along a linear path between two points
 * @param {object} left - point where interpolation is begun, of format {x: NUMBER, y: NUMBER}
 * @param {object} left - point where interpolation is ended, of format {x: NUMBER, y: NUMBER}
 * @param {number} numberOfPoints - the number of stops in the inteprolation
 * @returns {Array} array of interpolated points
 */
export function findEquidistantPoints(left, right, numberOfPoints) {
  if (numberOfPoints === 1) {
    return [left];
  }
  const slope = (right.y - left.y) / (right.x - left.x);
  const offset = left.y - slope * left.x;
  const xRange = right.x - left.x;
  const yRange = right.y - left.y;
  const slopeIsVertical = Number.isNaN(slope) || Math.abs(slope) === Infinity;

  const points = [];
  for (let i = 0; i < numberOfPoints; i++) {
    const fraction = i / (numberOfPoints - 1);
    const xVal = fraction * xRange + left.x;
    const newPoint = {
      x: slopeIsVertical ? left.x : xVal,
      y: slopeIsVertical ? fraction * yRange + left.y : (slope * xVal + offset)
    };
    points.push(newPoint);
  }
  return points;
}

/** Compute the area for a polygon with no holes
 * @param {array} points - array of objects formatted {x, y}
 * @returns {Number} the computed area
 */
export function area(points) {
  // double the signed area for the polygon
  // TODO THIS FUNCTION IS LIABLE TO NUMERICAL DISTORTION, WATCHOUT
  const segmentSum = points
  .reduce((acc, row, index) => {
    const nextRow = points[(index + 1) % points.length];
    return acc + (row.x * nextRow.y - nextRow.x * row.y);
  }, 0);
  return 0.5 * Math.abs(segmentSum);
}

/** Helper function for partitionTriangle, generates constants
 * @param {object} v0 - left vertex of triangle
 * @param {object} v1 - right vertex of triangle
 * @param {number} area  - the value associated with this subsection
 * @returns {object} the relavant constants
 */
function getConsts(v0, v1, area) {
  const A = v1.x - v0.x;
  const B = v1.y - v0.y;
  const C = 2 * area - pointAdd(v0, v1);
  const Cm = -2 * area - pointAdd(v0, v1);
  // const C = 2 * area + A * v0.y - B * v0.x;
  // const Cm = -2 * area + A * v0.y - B * v0.x;
  return {A, B, C, Cm};
}

/** Parition a triangle into 3 unequally weight sections
 * ADD EXPLANATION OF PICKING ALGORITHM
 * having generated these lines, we drop them into a 2x2 matrix
 * and take the inverse to solve the system, which has as it's solution
 * the partition point p that we sought
 * @param {array} points - points {x, y} defining edge of triangle
 * @param {object} areas - {alpha, beta, gamma}
 * @returns {object} the partitions for each of the areas
 */
export function partitionTriangle(points, {alpha, beta, gamma}) {
  const [pointA, pointB, pointC] = points;
  const alphas = getConsts(pointB, pointC, alpha);
  const gammas = getConsts(pointA, pointB, gamma);
  const knownArea = alpha + beta + gamma;

  let det = 1 / (gammas.A * alphas.B - gammas.B * alphas.A);
  det = !isFinite(det) ? 0 : det;
  // need to check all of the Plus Minus combos, bc of abs value
  const result = [
    {gammaC: gammas.C, alphaC: alphas.C},
    {gammaC: gammas.Cm, alphaC: alphas.C},
    {gammaC: gammas.C, alphaC: alphas.Cm},
    {gammaC: gammas.Cm, alphaC: alphas.Cm}
  ].reduce((solution, combo) => {
    const {gammaC, alphaC} = combo;
    const partition = {
      x: det * (alphas.A * gammaC - gammas.A * alphaC),
      y: det * (alphas.B * gammaC - gammas.B * alphaC)
    };

    const trialSolution = {
      alpha: [partition, pointB, pointC],
      beta: [pointA, pointC, partition],
      gamma: [pointA, partition, pointB]
    };

    const solutionArea = area(trialSolution.alpha) + area(trialSolution.beta) + area(trialSolution.gamma);
    const solutionScore = Math.abs(solutionArea - knownArea);
    return {
      score: solutionScore < solution.score ? solutionScore : solution.score,
      solution: solutionScore < solution.score ? trialSolution : solution.solution
    };
  }, {score: Infinity, solution: false});

  return {
    ...result.solution,
    partition: result.solution.alpha[0]
  };
}

function getConst(pp, targArea, print) {
  const points = pp;
  points.reverse();
  let sum = 0
  for (let i = 0; i < points.length - 1; i++) {
    sum += pointAdd(points[i], points[(i + 1)])
  }

  const aa = points[points.length - 1];
  const bb = points[0];
  const left = aa.x > bb.x ? aa : bb;
  const right = aa.x < bb.x ? aa : bb;

  if (print) {

    [-2 * targArea, 2 * targArea, 0, -8 * targArea].forEach(arrrr => {
      const slope = -(right.y - left.y) / (left.x - right.x);
      const intersect = (arrrr - sum) / (left.x - right.x);
      const evalLin = (point) => ({x: point.x, y: slope * point.x + intersect})
      // console.log(evalLin(left), evalLin(right))
    });
  }

  return {
    A: 2 * targArea - sum,
    Am: -8 * targArea - sum,
    B: left.x - right.x,
    C: right.y - left.y
  };
}

/** Generate a point that is a particular fraction between two other points
 * @param {object} left - point {x, y} to interpolate from
 * @param {object} right - point {x, y} to interpolate to
 * @param {number} fraction - fraction of the distance between them to interpolate
 * @returns {object} the partitions for each of the areas
 */
export function fractionalInterpolation(left, right, fraction) {
  const slope = (right.y - left.y) / (right.x - left.x);
  const offset = left.y - slope * left.x;
  const xRange = right.x - left.x;
  const yRange = right.y - left.y;
  const slopeIsVertical = Number.isNaN(slope) || Math.abs(slope) === Infinity;

  const xVal = fraction * xRange + left.x;
  return {
    x: slopeIsVertical ? left.x : xVal,
    y: slopeIsVertical ? fraction * yRange + left.y : (slope * xVal + offset)
  };
}

/** Parition a quadrangle into 3 unequally weight sections
 * @param {array} points - points {x, y} defining edge of quadrangle
 * @param {object} areas - {alpha, beta, gamma}
 * @returns {object} the partitions for each of the areas
 */
export function partitionQuadrangle(points, intersectionPoints, {alpha, beta, gamma}, reorder) {
  // const [pointA, pointB, pointC, pointD, pointE] = points;
  // const totalArea = alpha + beta + gamma;
  // // TODO should this be 1?
  // const fractionalArea = totalArea > 0 ? (beta + gamma) / totalArea : 0;
  // // console.log('fractional area', fractionalArea)
  // const betaLeft = fractionalInterpolation(pointA, pointB, 1 - fractionalArea);
  // const gammaRight = fractionalInterpolation(pointE, pointD, 1 - fractionalArea);
  // // console.log(betaLeft, gammaRight)
  // // console.log(points)
  // // const [betaLeft, gammaRight] = intersectionPoints;
  //
  // // this use of max's assumes were right about everything else
  // // so watch out
  // const outerAlpha = area([gammaRight, pointE, pointA, betaLeft]);
  // const innerAlpha = Math.max(alpha - outerAlpha, 0);
  //
  // const outerGamma = area([pointC, pointD, gammaRight]);
  // const innerGamma = Math.max(gamma - outerGamma, 0);
  //
  // const outerBeta = area([betaLeft, pointB, pointC]);
  // const innerBeta = Math.max(beta - outerBeta, 0);
  //
  // const triangelPoints = reorder ? [pointC, gammaRight, betaLeft] : [pointC, betaLeft, gammaRight];
  //
  // const triangleSoln = partitionTriangle(
  //   triangelPoints,
  //   // [pointC, betaLeft, gammaRight],
  //   // the below is left over from finding the previous line
  //   // if you are using it, you are probably wrong
  //   // [pointC, gammaRight, betaLeft],
  //   // it is sometimes the case that these input need to be reverered???????
  //   {alpha: innerAlpha, beta: innerBeta, gamma: innerGamma}
  // );
  // // console.log({innerAlpha, innerBeta, innerGamma}, {alpha, beta, gamma})
  // // console.log('points used', JSON.stringify([pointC, betaLeft, gammaRight]))
  // // TODO MAGIC NUMBER!!
  // const partition = triangleSoln.partition;
  // // console.log('partition', partition)
  // return {
  //   alpha: [pointA, betaLeft, partition, gammaRight, pointE],
  //   beta: [partition, betaLeft, pointB, pointC],
  //   gamma: [pointC, pointD, gammaRight, partition]
  // };

  return canonSolution(points, intersectionPoints, {alpha, beta, gamma});
  // return altSolution(points, intersectionPoints, {alpha, beta, gamma});
}

function canonSolution(points, intersectionPoints, {alpha, beta, gamma}) {
  const [pointA, pointB, pointC, pointD, pointE] = points;
  const [betaLeft, gammaRight] = intersectionPoints;

  const outerAlpha = area([gammaRight, pointE, pointA, betaLeft]);
  const innerAlpha = alpha - outerAlpha;

  const outerGamma = area([pointC, pointD, gammaRight]);
  const innerGamma = gamma - outerGamma;

  const outerBeta = area([betaLeft, pointB, pointC]);
  const innerBeta = beta - outerBeta;

  const triangleSoln = partitionTriangle(
    [pointC, betaLeft, gammaRight],
    // [pointC, gammaRight, betaLeft],
    {alpha: innerAlpha, beta: innerBeta, gamma: innerGamma}
  );
  // console.log({innerAlpha, innerBeta, innerGamma}, {alpha, beta, gamma})
  // TODO MAGIC NUMBER!!
  const partition = triangleSoln.alpha[0];
  // console.log('partition', partition)
  return {
    alpha: [pointA, betaLeft, partition, gammaRight, pointE],
    beta: [partition, betaLeft, pointB, pointC],
    gamma: [pointC, pointD, gammaRight, partition]
  };
}

function altSolution(points, intersectionPoints, {alpha, beta, gamma}) {
  const [pointA, pointB, pointC, pointD, pointE] = points;
  const [betaLeft, gammaRight] = intersectionPoints;
  // const alphas = getConst([gammaRight, pointE, pointA, betaLeft], alpha);
  const alphas = getConst([gammaRight, pointD, pointC], gamma, true);
  const betas = getConst([betaLeft, pointB, pointC], beta);

  const knownArea = alpha + beta + gamma;
  // const knownArea = area(points);
  // console.log(knownArea, alpha, beta, gamma)
  const Amatrix = [[betas.C, betas.B], [alphas.C, alphas.B]];

  // need to check all of the Plus Minus combos, bc of abs value
  const result = [
    [betas.A, alphas.A],
    [betas.Am, alphas.A],
    [betas.A, alphas.Am],
    [betas.Am, alphas.Am]
  ].reduce((solution, Bvec) => {
    const partition = getXYforAB(Amatrix, Bvec);
    const trialSolution = {
      alpha: [pointA, betaLeft, partition, gammaRight, pointE],
      beta: [partition, betaLeft, pointB, pointC],
      gamma: [pointC, pointD, gammaRight, partition]
    };
    // console.log(alpha, area(trialSolution.alpha), beta, area(trialSolution.beta), gamma, area(trialSolution.gamma))
    const solutionArea = area(trialSolution.alpha) + area(trialSolution.beta) + area(trialSolution.gamma);
    // console.log('target', knownArea, solutionArea)
    const solutionScore = Math.abs(solutionArea - knownArea);
    // console.log(checkIfInside(points, partition))
    return {
      score: solutionScore < solution.score ? solutionScore : solution.score,
      solution: solutionScore < solution.score ? trialSolution : solution.solution
    };
  }, {score: Infinity, solution: false});

  return result.solution;
}

function getXYforAB(A, b) {
  const det = A[0][0] * A[1][1] - A[0][1] * A[1][0];
  return {
    x: (A[1][1] * b[0] - A[0][1] * b[1]) / det,
    y: (-A[1][0] * b[0] + A[0][0] * b[1]) / det
  };
}

/** Generate a polygon
 * @param {number} number - the number of sides in the polygon
 * @param {number} radius - the radius of the polygon
 * @param {object} offset - the offset of the polygon from zero, {x, y}
 * @returns {array} list of points in polygon
 */
export function generatePolygon(numberOfSides, radius, offset) {
  return new Array(numberOfSides).fill(0).map((point, index) => {
    const angle = Math.PI * 2 * index / numberOfSides + Math.PI / 3;
    return {x: radius * Math.cos(angle) + offset.x, y: radius * Math.sin(angle) + offset.y};
  });
}

export function deepCopyTable(table) {
  return table.reduce((acc, row) => acc.concat([row.slice(0)]), []);
}
