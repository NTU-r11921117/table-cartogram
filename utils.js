/**
 * Generate a number of equally spaced points along a linear path between two points
 * @param {object} left - point where interpolation is begun, of format {x: NUMBER, y: NUMBER}
 * @param {object} left - point where interpolation is ended, of format {x: NUMBER, y: NUMBER}
 * @param {number} numberOfPoints - the number of stops in the inteprolation
 * @returns {Array} array of interpolated points
 */
export function findEquidistantPoints(left, right, numberOfPoints) {
  if (numberOfPoints === 1) {
    return [left];
  }
  const slope = (right.y - left.y) / (right.x - left.x);
  const offset = left.y - slope * left.x;
  const xRange = right.x - left.x;
  const yRange = right.y - left.y;
  const slopeIsVertical = Number.isNaN(slope) || Math.abs(slope) === Infinity;

  const points = [];
  for (let i = 0; i < numberOfPoints; i++) {
    const fraction = i / (numberOfPoints - 1);
    const xVal = fraction * xRange + left.x;
    const newPoint = {
      x: slopeIsVertical ? left.x : xVal,
      y: slopeIsVertical ? fraction * yRange + left.y : (slope * xVal + offset)
    };
    points.push(newPoint);
  }
  return points;
}

/** Compute the area for a polygon with no holes
 * @param {array} points - array of objects formatted {x, y}
 * @returns {Number} the computed area
 */
export function area(points) {
  // double the signed area for the polygon
  // TODO THIS FUNCTION IS LIABLE TO NUMERICAL DISTORTION, WATCHOUT
  const segmentSum = points
  .reduce((acc, row, index) => {
    const nextRow = points[(index + 1) % points.length];
    return acc + (row.x * nextRow.y - nextRow.x * row.y);
  }, 0);
  return 0.5 * Math.abs(segmentSum);
}

/** Helper function for partitionTriangle, generates constants
 * @param {object} v0 - left vertex of triangle
 * @param {object} v1 - right vertex of triangle
 * @param {number} area  - the value associated with this subsection
 * @returns {object} the relavant constants
 */
function getConsts(v0, v1, area) {
  const A = v1.x - v0.x;
  const B = v1.y - v0.y;
  const C = 2 * area - pointAdd(v0, v1);
  const Cm = -2 * area - pointAdd(v0, v1);
  // const C = 2 * area + A * v0.y - B * v0.x;
  // const Cm = -2 * area + A * v0.y - B * v0.x;
  return {A, B, C, Cm};
}

/** Parition a triangle into 3 unequally weight sections
 * ADD EXPLANATION OF PICKING ALGORITHM
 * having generated these lines, we drop them into a 2x2 matrix
 * and take the inverse to solve the system, which has as it's solution
 * the partition point p that we sought
 * @param {array} points - points {x, y} defining edge of triangle
 * @param {object} areas - {alpha, beta, gamma}
 * @returns {object} the partitions for each of the areas
 */
export function partitionTriangle(points, {alpha, beta, gamma}) {
  const [pointA, pointB, pointC] = points;
  const alphas = getConsts(pointB, pointC, alpha);
  const gammas = getConsts(pointA, pointB, gamma);
  const knownArea = alpha + beta + gamma;

  let det = 1 / (gammas.A * alphas.B - gammas.B * alphas.A);
  det = !isFinite(det) ? 0 : det;
  // need to check all of the Plus Minus combos, bc of abs value
  const result = [
    {gammaC: gammas.C, alphaC: alphas.C},
    {gammaC: gammas.Cm, alphaC: alphas.C},
    {gammaC: gammas.C, alphaC: alphas.Cm},
    {gammaC: gammas.Cm, alphaC: alphas.Cm}
  ].reduce((solution, combo) => {
    const {gammaC, alphaC} = combo;
    const partition = {
      x: det * (alphas.A * gammaC - gammas.A * alphaC),
      y: det * (alphas.B * gammaC - gammas.B * alphaC)
    };

    const trialSolution = {
      alpha: [partition, pointB, pointC],
      beta: [pointA, pointC, partition],
      gamma: [pointA, partition, pointB]
    };

    const solutionArea = area(trialSolution.alpha) + area(trialSolution.beta) + area(trialSolution.gamma);
    const solutionScore = Math.abs(solutionArea - knownArea);
    return {
      score: solutionScore < solution.score ? solutionScore : solution.score,
      solution: solutionScore < solution.score ? trialSolution : solution.solution
    };
  }, {score: Infinity, solution: false});

  return result.solution;
}

const pointAdd = (l, r) => (l.x * r.y - r.x * l.y);
function getConst(pp, area, print) {
  const points = pp;
  points.reverse();
  let sum = 0
  for (let i = 0; i < points.length - 1; i++) {
    sum += pointAdd(points[i], points[(i + 1)])
  }

  const aa = points[points.length - 1];
  const bb = points[0];
  const left = aa.x > bb.x ? aa : bb;
  const right = aa.x < bb.x ? aa : bb;

  if (print) {

    [-2 * area, 2 * area, 0, -8 * area].forEach(arrrr => {
      const slope = -(right.y - left.y) / (left.x - right.x);
      const intersect = (arrrr - sum) / (left.x - right.x);
      const evalLin = (point) => ({x: point.x, y: slope * point.x + intersect})
    });
  }

  return {
    A: 2 * area - sum,
    Am: -8 * area - sum,
    B: left.x - right.x,
    C: right.y - left.y
  };
}

/** Parition a quadrangle into 3 unequally weight sections
 * @param {array} points - points {x, y} defining edge of quadrangle
 * @param {object} areas - {alpha, beta, gamma}
 * @returns {object} the partitions for each of the areas
 */
export function partitionQuadrangle(points, intersectionPoints, {alpha, beta, gamma}) {
  const [pointA, pointB, pointC, pointD, pointE] = points;
  const [betaLeft, gammaRight] = intersectionPoints;

  const outerAlpha = area([gammaRight, pointE, pointA, betaLeft]);
  const innerAlpha = alpha - outerAlpha;

  const outerGamma = area([pointC, pointD, gammaRight]);
  const innerGamma = gamma - outerGamma;

  const outerBeta = area([betaLeft, pointB, pointC]);
  const innerBeta = beta - outerBeta;

  const triangleSoln = partitionTriangle(
    [pointC, betaLeft, gammaRight],
    // [pointC, gammaRight, betaLeft],
    {alpha: innerAlpha, beta: innerBeta, gamma: innerGamma}
  );
  // console.log({alpha: innerAlpha, beta: innerBeta, gamma: innerGamma}, {alpha, beta,gamma})
  const partition = triangleSoln.alpha[0];
  return {
    alpha: [pointA, betaLeft, partition, gammaRight, pointE],
    beta: [partition, betaLeft, pointB, pointC],
    gamma: [pointC, pointD, gammaRight, partition]
  };

  //
  // // const alphas = getConst([gammaRight, pointE, pointA, betaLeft], alpha);
  // const alphas = getConst([gammaRight, pointD, pointC], gamma, true);
  // const betas = getConst([betaLeft, pointB, pointC], beta);
  //
  // const knownArea = alpha + beta + gamma;
  // // const knownArea = area(points);
  // const Amatrix = [[betas.C, betas.B], [alphas.C, alphas.B]];

  // need to check all of the Plus Minus combos, bc of abs value
  // const result = [
  //   [betas.A, alphas.A],
  //   [betas.Am, alphas.A],
  //   [betas.A, alphas.Am],
  //   [betas.Am, alphas.Am]
  // ].reduce((solution, Bvec) => {
  //   const partition = getXYforAB(Amatrix, Bvec);
  //   const trialSolution = {
  //     alpha: [pointA, betaLeft, partition, gammaRight, pointE],
  //     beta: [partition, betaLeft, pointB, pointC],
  //     gamma: [pointC, pointD, gammaRight, partition]
  //   };
  //   // console.log(alpha, area(trialSolution.alpha), beta, area(trialSolution.beta), gamma, area(trialSolution.gamma))
  //   const solutionArea = area(trialSolution.alpha) + area(trialSolution.beta) + area(trialSolution.gamma);
  //   // console.log('target', knownArea, solutionArea)
  //   const solutionScore = Math.abs(solutionArea - knownArea);
  //   // console.log(checkIfInside(points, partition))
  //   return {
  //     score: solutionScore < solution.score ? solutionScore : solution.score,
  //     solution: solutionScore < solution.score ? trialSolution : solution.solution
  //   }
  // }, {score: Infinity, solution: false});

  // return result.solution;
}

function getXYforAB(A, B) {
  const det = A[0][0] * A[1][1] - A[0][1] * A[1][0];
  return {
    x: (A[1][1] * B[0] - A[0][1] * B[1]) / det,
    y: (-A[1][0] * B[0] + A[0][0] * B[1]) / det
  };
}

/** Round a value
 * @param {number} number - the number to be rounded
 * @param {number} precision - the precision of the rounding, expressed in powers of 10
 * @returns {number} the rounded value
 */
export function round(number, precision = Math.pow(10, 12)) {
  return Math.floor(number * precision) / precision;
}

/** Generate a polygon
 * @param {number} number - the number of sides in the polygon
 * @param {number} radius - the radius of the polygon
 * @param {object} offset - the offset of the polygon from zero, {x, y}
 * @returns {array} list of points in polygon
 */
export function generatePolygon(numberOfSides, radius, offset) {
  return new Array(numberOfSides).fill(0).map((point, index) => {
    const angle = Math.PI * 2 * index / numberOfSides + Math.PI / 3;
    return {x: radius * Math.cos(angle) + offset.x, y: radius * Math.sin(angle) + offset.y};
  });
}

/** Compute geometric center of a polygon
 * @param {array} points - list of points in polygon, present as {x, y}
 * @returns {object} center point of polygon
 */
export function geoCenter(points) {
  const sum = points.reduce((center, row) => {
    return {x: center.x + row.x, y: center.y + row.y};
  }, {x: 0, y: 0});
  return {x: sum.x / points.length, y: sum.y / points.length};
}
