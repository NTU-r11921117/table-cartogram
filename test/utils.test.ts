import {area, generatePolygon, fusePolygons} from '../src/utils';
import SavedBirdStrikes from './saved-bird-strikes.json';

test('area', () => {
  const SQUARE = [
    {x: 10, y: 10},
    {x: 20, y: 10},
    {x: 20, y: 20},
    {x: 10, y: 20},
  ];
  expect(area(SQUARE)).toEqual(100);

  const TRIANGLE = [
    {x: 10, y: 10},
    {x: 20, y: 10},
    {x: 20, y: 20},
  ];
  expect(area(TRIANGLE)).toEqual(50);

  const POLYGON = generatePolygon(5, 10, {x: 10, y: 15});
  expect(area(POLYGON)).toEqual(237.7641290737884);
});

test('fusePolygons', () => {
  const results = fusePolygons(SavedBirdStrikes, (d) => d.data.name);
  expect(results).toMatchSnapshot();
});
