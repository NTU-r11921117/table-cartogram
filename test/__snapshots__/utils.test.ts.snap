// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fusePolygons 1`] = `
Array [
  Object {
    "key": "Midwest",
    "vertices": Array [
      Object {
        "x": 0.22069152777643303,
        "y": 0.07384897947723086,
      },
      Object {
        "x": 0.2089601071841638,
        "y": 0.027034695775387203,
      },
      Object {
        "x": 0.18742933077935856,
        "y": 0.13217174603531479,
      },
      Object {
        "x": 0.20475207028770973,
        "y": 0.19502318026182358,
      },
      Object {
        "x": 0.21770495090460693,
        "y": 0.23911370833694068,
      },
      Object {
        "x": 0.14084096957945416,
        "y": 0.3,
      },
      Object {
        "x": 0,
        "y": 0.3,
      },
      Object {
        "x": 0,
        "y": 0,
      },
      Object {
        "x": 0.27735094126015575,
        "y": 0,
      },
    ],
  },
  Object {
    "key": "South",
    "vertices": Array [
      Object {
        "x": 0.6149107177284888,
        "y": 0.21981997812709772,
      },
      Object {
        "x": 0.580545301695947,
        "y": 0.16546780446830217,
      },
      Object {
        "x": 0.6105954017579522,
        "y": 0.09726417525665841,
      },
      Object {
        "x": 0.5791045681599444,
        "y": 0.0436604285195686,
      },
      Object {
        "x": 0.5689087495205509,
        "y": 0,
      },
      Object {
        "x": 0.27735094126015575,
        "y": 0,
      },
      Object {
        "x": 0.22069152777643303,
        "y": 0.07384897947723086,
      },
      Object {
        "x": 0.25847661265936483,
        "y": 0.1169815046936408,
      },
      Object {
        "x": 0.18742933077935856,
        "y": 0.13217174603531479,
      },
      Object {
        "x": 0.20475207028770973,
        "y": 0.19502318026182358,
      },
      Object {
        "x": 0.21770495090460693,
        "y": 0.23911370833694068,
      },
      Object {
        "x": 0.14084096957945416,
        "y": 0.3,
      },
      Object {
        "x": 0.5419205447339661,
        "y": 0.3,
      },
    ],
  },
  Object {
    "key": "West",
    "vertices": Array [
      Object {
        "x": 0.8841217298039928,
        "y": 0,
      },
      Object {
        "x": 0.5689087495205509,
        "y": 0,
      },
      Object {
        "x": 0.5791045681599444,
        "y": 0.0436604285195686,
      },
      Object {
        "x": 0.6105954017579522,
        "y": 0.09726417525665841,
      },
      Object {
        "x": 0.580545301695947,
        "y": 0.16546780446830217,
      },
      Object {
        "x": 0.6149107177284888,
        "y": 0.21981997812709772,
      },
      Object {
        "x": 0.5419205447339661,
        "y": 0.3,
      },
      Object {
        "x": 0.7405234764147632,
        "y": 0.3,
      },
      Object {
        "x": 0.7917170685482705,
        "y": 0.2560530434164313,
      },
      Object {
        "x": 0.780719453263419,
        "y": 0.17044456136426042,
      },
      Object {
        "x": 0.8006878406527494,
        "y": 0.09123489461020409,
      },
      Object {
        "x": 0.8582487193618223,
        "y": 0.12512749493592018,
      },
      Object {
        "x": 0.8160673650588451,
        "y": 0.07078434761770484,
      },
    ],
  },
  Object {
    "key": "Northeast",
    "vertices": Array [
      Object {
        "x": 1,
        "y": 0.21731063089827807,
      },
      Object {
        "x": 1,
        "y": 0,
      },
      Object {
        "x": 0.8841217298039928,
        "y": 0,
      },
      Object {
        "x": 0.8160673650588451,
        "y": 0.07078434761770484,
      },
      Object {
        "x": 0.8582487193618223,
        "y": 0.12512749493592018,
      },
      Object {
        "x": 0.8006878406527494,
        "y": 0.09123489461020409,
      },
      Object {
        "x": 0.780719453263419,
        "y": 0.17044456136426042,
      },
      Object {
        "x": 0.7917170685482705,
        "y": 0.2560530434164313,
      },
      Object {
        "x": 0.7405234764147632,
        "y": 0.3,
      },
      Object {
        "x": 0.9320327659099618,
        "y": 0.3,
      },
      Object {
        "x": 0.8929245664331396,
        "y": 0.17062989884011562,
      },
    ],
  },
  Object {
    "key": "Canada",
    "vertices": Array [
      Object {
        "x": 0.18742933077935856,
        "y": 0.13217174603531479,
      },
      Object {
        "x": 0.2089601071841638,
        "y": 0.027034695775387203,
      },
      Object {
        "x": 0.22069152777643303,
        "y": 0.07384897947723086,
      },
      Object {
        "x": 0.25847661265936483,
        "y": 0.1169815046936408,
      },
    ],
  },
  Object {
    "key": "US Islands",
    "vertices": Array [
      Object {
        "x": 1,
        "y": 0.3,
      },
      Object {
        "x": 1,
        "y": 0.21731063089827807,
      },
      Object {
        "x": 0.8929245664331396,
        "y": 0.17062989884011562,
      },
      Object {
        "x": 0.9320327659099618,
        "y": 0.3,
      },
    ],
  },
]
`;
